cohttp-lwt/src/server.ml

```diff
@@ -152,9 +153,28 @@ module Make (IO : S.IO) = struct
   let callback spec io_id ic oc =
     let conn_id = Cohttp.Connection.create () in
     let conn_closed () = spec.conn_closed (io_id, conn_id) in
+    let handle () = handle_client ic oc (io_id, conn_id) spec in
+    let is_conn_closed _sleep_fn =
+      (* Without a sleep function we cannot safely loop waiting for EOF *)
+      match spec.sleep_fn with
+      | None -> fst (Lwt.task ()) (* wait to be cancelled *)
+      | Some sleep_fn ->
+          IO.wait_eof_or_closed io_id ic sleep_fn >>= fun () ->
+          Log.debug (fun m ->
+              m "Client closed the connection, got EOF for %s"
+                (Cohttp.Connection.to_string conn_id));
+          Lwt.return Call_conn_closed
+    in
     Lwt.catch
       (fun () ->
-        IO.catch (fun () -> handle_client ic oc (io_id, conn_id) spec)
+        IO.catch (fun () ->
+            Lwt.pick [ handle (); is_conn_closed spec.sleep_fn ] >>= function
+            | Call_conn_closed ->
+                conn_closed ();
+                Lwt.return_unit
+            | Call_conn_closed_and_drain body ->
+                conn_closed ();
+                Body.drain_body body >>= fun () -> Lwt.return_unit)
```

Comment:
```suggestion
                Body.drain_body body)
```

Reply:
Removed in [fixup1](https://github.com/gabrielmoise17/ocaml-cohttp/pull/3/commits/b9a2a286f87995beb19446b72c55760e89b5e5a9)